### DISCUSSION TITLE
Confusing "Your project has pages but the `<NuxtPage />` component has not been used." error

### DISCUSSION THREAD

**Discussion by LLCampos** [nuxt/nuxt#25808] [Questions]
Hi :)

After upgrading from Nuxt 3.7 to 3.10 I started getting a bunch of these warns:

```
[nuxt] Your project has pages but the `<NuxtPage />` component has not been used. You might be using the `<RouterView />` component instead, which will not work correctly in Nuxt. You can set `pages: false` in `nuxt.config` if you do not wish to use the Nuxt `vue-router` integration.
````

The thing is, I'm not using `<RouterView />`. I'm using [file-based routing](https://nuxt.com/docs/guide/directory-structure/pages), not even having an `app.vue`.

What could I be missing?

**Comment by TheAlexLichter** [comment]
Hard to tell spontaneously. Feel free to share some code for more help

**Reply by LLCampos** [reply]
Its one of those situations where the codebase is too big for me to actually being able to easily share a reproduction. But I'll see what I can do.

**Reply by TheAlexLichter** [reply]
@LLCampos Understandable, but it'd be even harder for people trying to debug to go through a whole project 🙈 

Hope you can figure out a minimal repro 👍

**Comment by lna1989** [comment]
Most likely you have created the pages directory, but you still have the default app.vue file, which does not use the NuxtPage component.
Change app.vue from:
```html
<template>
  <div>
    <NuxtWelcome />
  </div>
</template>
```
to:
```html
<template>
  <div>
    <NuxtPage />
  </div>
</template>
```

If you are not using the app.vue, then most likely it takes the standard one from the nuxt module (but this is not accurate)

**Reply by LLCampos** [reply]
Hi :) Thanks for the answer. As written in the original question, I don't have an `app.vue`.

**Reply by lna1989** [reply]
If you delete app.vue file,  most likely, the standard app.vue is used from the nuxt module and it also lacks NuxtPage component in template.

**Reply by TheAlexLichter** [reply]
When using `pages` then a minimal app.vue should be used which should not cause the warning.

**Comment by Ambrise** [comment]
Hello, I have the same problem since Nuxt 3.10 :)

**Reply by TheAlexLichter** [reply]
>  Feel free to share some code for more help

**Comment by LLCampos** [comment]
This is fixed in [3.10.2](https://github.com/nuxt/nuxt/releases/tag/v3.10.2) :)

**Comment by ZackNew** [comment]
i had the same issue but adding <slot /> in my /layouts/default.vue fixed it.

**Comment by anthonyfranc** [comment]
For anyone else, the correct answer is using the following inside your layouts 

`<slot /> `

**Reply by farhaddr90** [reply]
Thanks I used <slot/> and it worked.

**Reply by ra-dave** [reply]
> For anyone else, the correct answer is using the following inside your layouts
> 
> `<slot /> `

Neither `<slot />` nor `<NuxtPage />` in my default.vue layout makes the warning go away.  Using Nuxt 3.13.2.

Is there any problem if this warning exists in production?  My site is working fine.

**Reply by rowild** [reply]
I have the same warnings in Devtools popping up. The website works, though. This is strange...

**Comment by skoenfaelt** [comment]
`app.vue`
```
<template>
  <div>
    <NuxtLayout>
      <NuxtPage />
    </NuxtLayout>
  </div>
</template>
```

and `<slot />` in all layouts. still having this output on nuxt 3.14.1592

**Comment by JonasKabalo** [comment]
I am on version `3.14.1592` as well and see the error comes up with `<slot />` in `layouts/default.vue` 🙁

**Reply by JonasKabalo** [reply]
One thing tho is that i have a statement on my slot like that `<slot v-if="" />`.
I don't know what I need to do to make the warning go away, but I need some logic before I display the content! if i removed it, the warning go away, but need some logic before so i dunno what really to do to make it go away

**Reply by TheAlexLichter** [reply]
if you have a v-if there and things work as expected, you can ignore the warning.

**Comment by Ahmard** [comment]
This warning can also appear when you're slot is conditional.
Below reproduces this warning for me.
```vue
<SplashScreen v-if="isLoading" />
<slot v-else />
```

**Comment by ArkBeam-Amar** [comment]
I get this warning too, because I am conditionally rendering `<NuxtPage />` based on auth state. My auth server is an external auth provider (propelauth). So on page load, it takes a second to validate refresh token from the external auth server (which is not my own nitro server).

```vue
<template>
  <NuxtLoadingIndicator />
  <div v-if="auth">
    <NuxtLayout>
      <NuxtPage />
    </NuxtLayout>
  </div>
  <UContainer v-else>
    <LoginCard class="mt-24 max-w-md mx-auto" />
  </UContainer>
  <UNotifications />
</template>

<script setup lang="ts">
const auth = useAuthState();
</script>

<style scoped></style>
```

This was the easiest architecture for me without having to manage layouts for auth and non authed states, or managing redirects.

But as Alex mentioned, since it works as expected, I am choosing to ignore the warning for now.

**Comment by Psycarlo** [comment]
I add this issue and my problem was not adding a `slot` to my `auth` layout.
Every layout should have a `slot`.

**Comment by PetyXbron** [comment]
Yeah, for me, I just fixed it by updating `nuxt` module to the latest version.

**Comment by Kealtie** [comment]
I've had the same issue, this is probably because one of your layouts has a NuxtPage which is conditionally rendered

` <div v-if="!pending" class="page-content">
            <NuxtPage />
        </div>`

**Comment by kimDanao** [comment]
check app.vue and layouts/default.vue
if the app.vue and layouts/default.vue uses <NuxtPage /> then try changing the one on layouts/default.vue to slot

**Comment by VectorChi** [comment]
To solve the conditional rendering issue, I used.

/layouts/default

```
<LoadingPage v-if="isLoading" />
<div v-show="!isLoading">
  <slot/>
</div>
```

**Comment by JonasKabalo** [comment]
OK, just sharing my thoughts! "This error occurs because Nuxt is still looking for pages in the `pages/` directory,". I've just tried deleting my pages directory and created an `app.vue` with this: (And I do have a `default/layout` of course!)

```
<NuxtLayout name="default">
 <NuxtPage />
</NuxtLayout>
```

And I've put all the contents of the `pages/` directory into a directory called `routes/` and boom the warning disappears !!!!
Now explain to me what's going on, because I still get the warning with the latest version!! What the heck 🤬

**Reply by JonasKabalo** [reply]
Looks like `pages/` folder might be the problem here 🤔

**Comment by yassine-safraoui** [comment]
I wanted to share my experience with this warning. In my layout, I centralize authentication logic: while the session is loading, I show a loader and **don't render the slot**; once loaded, I either redirect to login or show the slot if authenticated. Example:

```vue
<template>
  <main>
    <div v-if="session.isPending || session.isRefetching">Loading...</div>
    <slot v-else="session.data" />
  </main>
</template>
<script setup>
const session = authClient.useSession()
const { path } = useRoute()
watch(session, s => {
  if (s.isPending || s.isRefetching) return
  if (!s.data && path !== '/login') navigateTo('/login', { replace: true })
}, { immediate: true })
</script>
```

Using `v-show` on a wrapper removes the warning, but the slot’s component still executes on the server while auth is loading, so it doesn’t actually gate rendering. This means I’d need to check auth in every page, which I’m trying to avoid. Middleware solves this but delays the page and also doesn’t allow showing a loader.

I think layouts that gate rendering based on async auth are a valid use case, but current Nuxt behavior makes this tricky. I could just ignore the warning for this use case but that isn't clean.
Would appreciate any best practices or guidance!

**Comment by kayheypark** [comment]
❌ In my case, the issue occured by "v-if":
```javascript
<slot v-if="!loading" />
```
<img width="840" height="107" alt="image" src="https://github.com/user-attachments/assets/c78dad12-6ab8-49cc-8e09-869d3837b023" />




✅ I proved the issue:

```javascript
<template>
  <div :class="{ hidden: loading }">
    <slot />
  </div>
</template>

<style scoped>
  .hidden {
    visibility: hidden;
  }
</style>
```
<img width="839" height="290" alt="image" src="https://github.com/user-attachments/assets/efed5cca-8026-4d3c-9516-be9c5fea6cee" />


nuxt 4.2.1

**Comment by baoanhng** [comment]
The issue happens the most when the main <slot /> is rendered with conditions.

**Comment by sukumaar** [comment]
I am following every suggestions mentioned in this thread in the project with nuxt 4.3.x but still seeing that issue. This seems false warning and a little bit irritating.

**Reply by sukumaar** [reply]
It seems this is getting called erroneously https://github.com/nuxt/nuxt/blob/f2ef4d4754b8a0e8b8a6e14f320db8b820762a09/packages/nuxt/src/pages/runtime/plugins/check-if-page-unused.ts#L11

**Comment by dbkup** [comment]
I got the same error today starting a new project (Nuxt 4.2.2 (with Nitro 2.13.1, Vite 7.3.1 and Vue 3.5.26)).

Project init with `npm create nuxt@latest nuxt-test` . I used a `minimal` template, and only two modules:
<img width="961" height="126" alt="image" src="https://github.com/user-attachments/assets/703423ee-1628-4581-9786-77f2ee3f3afa" />
Then I added a new page with `npx nuxt add page about`. Then run `npx nuxt dev`. And get the warning:
<img width="969" height="319" alt="image" src="https://github.com/user-attachments/assets/6f23c0ad-9dcc-4385-a16d-ba8762befa5a" />

As someone trying out Nuxt it's odd to get a warning by using the suggested tooling. I can fix it by removing the `app.vue` file.

Edit: add last paragraph

**Reply by franklin-tina** [reply]
From the [Nuxt docs](https://nuxt.com/docs/4.x/directory-structure/app/app)
<img width="717" height="99" alt="Screenshot 2026-01-17 at 19 37 06" src="https://github.com/user-attachments/assets/3521bdde-f724-48a8-8d44-9eae65df272a" />
So yes, removing `app.vue` will fix the issue


If you want to keep `app.vue` and use pages, make sure:
1. You have a `pages/index.vue` file
2. Your `app.vue` should look something like
```vue
<template>
  <div>
    <NuxtPage />
    <NuxtRouteAnnouncer />
  </div>
</template>
```

Edit: `<NuxtPage />` is the important piece here

**Reply by dbkup** [reply]
Thanks, I gathered both possible fixes to resolve the warning from this issue. 
For me it discouraging when you're told/given the option to add pages by using the CLI and when you do you get a warning. As someone who's created my first nuxt project today I saw multiple warnings and that makes me uncomfortable.
I also don't understand what the difference between app and pages is and I don't know which one is more for me 😄 
Perhaps that notice can be updated to expand: when adding pages you may have a warning which can be resolved by: option1 and option2.

**Comment by alanscodelog** [comment]
It would be nice to be able to ignore this error on certain conditions. I ran into this gating the main app pages from rendering when the app is active on another tab. Doing it from the layout seemed quite clean but produces this annoying error.
